fabtests/benchmarks/rdm_rpc_bw.c

```diff
@@ -0,0 +1,658 @@
+			    MIN(opts.window_size, rpc_limit - rpc_count)) {
+				ret = rpc_desc_server_post_recv(msg_event.op_context);
+				ft_assert(ret == FI_SUCCESS);
+			}
+		} else {
+			ret = rpc_desc_server_process_rma_response(msg_event.op_context);
+			ft_assert(ret == FI_SUCCESS);
+		}
+	}
+}
+
+static void rpc_process_rx_cq(size_t rpc_limit)
+{
+	struct fi_cq_tagged_entry msg_event;
+	int ret;
+
+	ret = fi_cq_read(rxcq, &msg_event, 1);
+	if (ret == -FI_EAGAIN)
+		return;
+
+	ft_assert(ret == 1);
+
+	if (rpc_client) {
+		ft_assert(msg_event.flags == (FI_TAGGED | FI_RECV));
+
+		ret = rpc_desc_client_process_recv(msg_event.op_context);
+		ft_assert(ret == FI_SUCCESS);
+
+		/* Repost the RPC request if the limit has not been reached. */
+		if (rpc_inflight <
+		    MIN(opts.window_size, rpc_limit - rpc_count)) {
+			ret = rpc_desc_client_send_req(msg_event.op_context);
+			ft_assert(ret == FI_SUCCESS);
+		}
+	} else {
+		ft_assert(msg_event.flags == (FI_MSG | FI_RECV));
+
+		ret = rpc_desc_server_process_req(msg_event.op_context);
+		ft_assert(ret == FI_SUCCESS);
+	}
+}
+
+static void rpc_server_run_benchmark(size_t rpc_limit)
+{
+	int ret;
+	size_t i;
+	struct rpc_desc *desc;
+
+	rpc_count = 0;
+	rpc_byte_count = 0;
+
+	ft_start();
+
+	/* Prepost receive buffers. */
+	for (i = 0; i < MIN(opts.window_size, rpc_limit); i++) {
+		desc = descs + i;
+
+		ret = rpc_desc_server_post_recv(desc);
+		ft_assert(ret == FI_SUCCESS);
+	}
+
+	do {
+		rpc_process_tx_cq(rpc_limit);
+		rpc_process_rx_cq(rpc_limit);
+	} while (rpc_count < rpc_limit);
+
+	ft_stop();
+}
+
+static void rpc_server_run(void)
+{
+	size_t i;
+	int ret;
+	struct rpc_desc *desc;
+
+	/* Register payload buffer for the duration of the test. */
+	for (i = 0; i < opts.window_size; i++) {
+		desc = descs + i;
+
+		ret = rpc_desc_reg_payload_mr(desc);
+		ft_assert(ret == FI_SUCCESS);
+	}
+
+	ft_sync();
+	rpc_server_run_benchmark(opts.warmup_iterations);
+
+	ft_sync();
+	rpc_server_run_benchmark(opts.iterations);
+
+	/* Deregister payload buffers with test completed of the test. */
+	for (i = 0; i < opts.window_size; i++) {
+		desc = descs + i;
+		FT_CLOSE_FID(desc->payload_mr);
+	}
+}
+
+static void rpc_client_run_benchmark(size_t rpc_limit)
+{
+	int ret;
+	size_t i;
+	struct rpc_desc *desc;
+
+	rpc_count = 0;
+	rpc_byte_count = 0;
+
+	ft_start();
+
+	/* Post RPC requests. */
+	for (i = 0; i < MIN(opts.window_size, rpc_limit); i++) {
+		desc = descs + i;
+
+		ret = rpc_desc_client_send_req(desc);
+		ft_assert(ret == FI_SUCCESS);
+	}
+
+	do {
+		rpc_process_tx_cq(rpc_limit);
+		rpc_process_rx_cq(rpc_limit);
+	} while (rpc_count < rpc_limit);
+
+	ft_stop();
+}
+
+static void rpc_client_run(void)
+{
+	ft_sync();
+	rpc_client_run_benchmark(opts.warmup_iterations);
+
+	ft_sync();
+	rpc_client_run_benchmark(opts.iterations);
+}
+
+static char *test_names[2][2] = {
+	{
+		"Server RPC Read Response",
+		"Server RPC Write Response",
+	},
+	{
+		"Client RPC Read Request",
+		"Client RPC Write Request",
+	},
+};
```

Comment:
This looks like it must be aligned with enum rpc_op_type values.  Please locate closer to that definition or use the enum values to initialize the array.